src/Interpreters/TransactionVersionMetadata.cpp

```diff
@@ -0,0 +1,431 @@
+            creation_csn.store(creation, std::memory_order_relaxed);
+        else
+            return false;
+    }
+
+    /// Part is probably visible for some transactions (part is too new or the oldest snapshot is too old)
+    if (oldest_snapshot_version < creation)
+        return false;
+
+    TIDHash removal_lock = removal_tid_lock.load(std::memory_order_relaxed);
+    /// Part is active
+    if (!removal_lock)
+        return false;
+
+    CSN removal = removal_csn.load(std::memory_order_relaxed);
+    if (!removal)
+    {
+        /// Part removal is not committed yet
+        removal = getCSNAndAssert(removal_lock, removal_csn);
+        if (removal)
+            removal_csn.store(removal, std::memory_order_relaxed);
+        else
+            return false;
+    }
+
+    /// We can safely remove part if all running transactions were started after part removal was committed
+    return removal <= oldest_snapshot_version;
+}
+
+#define CREATION_TID_STR "creation_tid: "
+#define CREATION_CSN_STR "creation_csn: "
+#define REMOVAL_TID_STR  "removal_tid:  "
+#define REMOVAL_CSN_STR  "removal_csn:  "
+
+
+void VersionMetadata::writeCSN(WriteBuffer & buf, WhichCSN which_csn, bool internal /* = false*/) const
+{
+    if (which_csn == CREATION)
+    {
+        if (CSN creation = creation_csn.load())
+        {
+            writeCString("\n" CREATION_CSN_STR, buf);
+            writeText(creation, buf);
+        }
+        else if (!internal)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "writeCSN called for creation_csn = 0, it's a bug");
+    }
+    else /// if (which_csn == REMOVAL)
+    {
+        if (CSN removal = removal_csn.load())
+        {
+            writeCString("\n" REMOVAL_CSN_STR, buf);
+            writeText(removal, buf);
+        }
+        else if (!internal)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "writeCSN called for removal_csn = 0, it's a bug");
+    }
+}
+
+void VersionMetadata::writeRemovalTID(WriteBuffer & buf, bool clear) const
+{
+    writeCString("\n" REMOVAL_TID_STR, buf);
+    if (clear)
+        TransactionID::write(Tx::EmptyTID, buf);
+    else
+        TransactionID::write(removal_tid, buf);
+}
+
+void VersionMetadata::write(WriteBuffer & buf) const
+{
+    writeCString("version: 1", buf);
+    writeCString("\n" CREATION_TID_STR, buf);
+    TransactionID::write(creation_tid, buf);
+    writeCSN(buf, CREATION, /* internal */ true);
+
+    if (removal_tid_lock)
+    {
+        assert(!removal_tid.isEmpty());
+        assert(removal_tid.getHash() == removal_tid_lock);
+        writeRemovalTID(buf);
+        writeCSN(buf, REMOVAL, /* internal */ true);
+    }
+}
+
+void VersionMetadata::read(ReadBuffer & buf)
+{
+    constexpr size_t size = sizeof(CREATION_TID_STR) - 1;
+    static_assert(sizeof(CREATION_CSN_STR) - 1 == size);
+    static_assert(sizeof(REMOVAL_TID_STR) - 1 == size);
+    static_assert(sizeof(REMOVAL_CSN_STR) - 1 == size);
+
+    assertString("version: 1", buf);
```

Comment:
Maybe parse it and give more clear exceptions?

Reply:
It will fail with clear message like `Cannot parse input: expected "version: 1" before "version: 2"` if version is different